services/subscriber/service.go

```diff
@@ -25,11 +27,66 @@ const (
 	statWriteFailures  = "writeFailures"
 )
 
+// WriteRequest is a parsed write request.
+type WriteRequest struct {
+	Database        string
+	RetentionPolicy string
+	// lineProtocol must be valid newline-separated line protocol.
+	lineProtocol []byte
+	// pointOffsets gives the starting index within lineProtocol of each point,
+	// for splitting batches if required.
+	pointOffsets []int
+}
+
+func NewWriteRequest(r *coordinator.WritePointsRequest, log *zap.Logger) (wr WriteRequest, numInvalid int64) {
+	log = log.With(zap.String("database", r.Database), zap.String("retention_policy", r.RetentionPolicy))
+	// Pre-allocate at least smallPointSize bytes per point.
+	const smallPointSize = 10
+	writeReq := WriteRequest{
+		Database:        r.Database,
+		RetentionPolicy: r.RetentionPolicy,
+		pointOffsets:    make([]int, 0, len(r.Points)),
+		lineProtocol:    make([]byte, 0, len(r.Points)*smallPointSize),
+	}
+	numInvalid = 0
+	for _, p := range r.Points {
+		if err := models.ValidPointStrings(p); err != nil {
+			log.Debug("discarding point", zap.Error(err))
+			numInvalid++
+			continue
+		}
+		writeReq.pointOffsets = append(writeReq.pointOffsets, len(writeReq.lineProtocol))
```

Comment:
Can you add a comment above these lines calling out that the ordering is required for the offset-logic to work?